### PR TITLE
Revert "[Core] Replace routing replay with device cache and async D2H pipeline" (#39917)

### PR DESCRIPTION
**Pull request by aoshen02**
## Summary

This reverts merge commit `8189a15914ca48461acf106f126c58ef7e41c9ee` (#39917). Net diff: -1495 / +733 across 15 files, including the deletion of `docs/training/routed_experts_replay.md` and a near-full rewrite of `vllm/model_executor/layers/fused_moe/routed_experts_capturer.py` back to the pre-39917 (SharedMemory-based) design.
So both because the author from nemo RL has some urgent releases so might not be able to support prefix caching and other features very soon, so we all agree to use #39568 and revert # 39917 for now. Thanks for the effort from both side again. cc @TomerBN-Nvidia

### CODE REVIEW

File: vllm/model_executor/layers/fused_moe/layer.py [hunk 1]
```diff
@@ -294,10 +291,6 @@ def __init__(
     ):
         super().__init__()
 
-        # Assign unique layer ID for routing replay buffer binding.
-        self.moe_layer_id = FusedMoE._next_moe_layer_id
-        FusedMoE._next_moe_layer_id += 1
-
         if params_dtype is None:
             params_dtype = torch.get_default_dtype()
         self.params_dtype = params_dtype
```
[the remark this answers is not in the document]
Comment (aoshen02) [reply vllm/model_executor/layers/fused_moe/layer.py:150]: Verified false alarm. `FusedMoE.layer_id` is a `@property` defined at `vllm/model_executor/layers/fused_moe/layer.py:484` (post-merge state in this branch):

```python
@property
def layer_id(self):
    from vllm.model_executor.models.utils import extract_layer_index
    return extract_layer_index(self.layer_name)
```

`module.layer_id` resolves to this property at runtime; no `AttributeError`. The `_next_moe_layer_id` auto-increment removed by the revert was added by #39917 itself — pre-#39917 code already used the layer-name-derived `layer_id` property and worked.

File: vllm/v1/worker/gpu_model_runner.py [hunk 2]
```diff
@@ -6996,44 +6970,45 @@ def init_routed_experts_capturer(self):
...
-        )
+        for module in self.compilation_config.static_forward_context.values():
+            if isinstance(module, FusedMoE) and isinstance(module.router, BaseRouter):
+                layer_id = module.layer_id
+
+                def _capture_fn(topk_ids, _layer_id=layer_id, _capturer=capturer):
+                    _capturer.capture(_layer_id, topk_ids)
 
-        bind_routing_capture_to_model(self.model)
+                module.router.set_capture_fn(_capture_fn)
```
[the remark this answers is not in the document]
Comment (aoshen02) [reply vllm/v1/worker/gpu_model_runner.py:7094]: Verified false alarm. The capturer buffer is sized by **total transformer layer count**, not MoE-layer count, in the pre-#39917 design. See `vllm/model_executor/layers/fused_moe/routed_experts_capturer.py` post-revert:

```python
num_layers = hf_config.num_hidden_layers  # line 147
self._device_buffer = torch.zeros((max_tokens, num_layers, ...), ...)  # line 151
```

So indexing the buffer with `module.layer_id` (global transformer layer index, 0..num_hidden_layers-1) is consistent with the dimension. Non-MoE layers reserve unused slots in the second axis — wastes a few KB but is functionally correct.

The MoE-specific `moe_layer_id` auto-increment was introduced *by* #39917 (along with a MoE-sized buffer) and is correctly reverted.

**Review by ywang96 — APPROVED** [review]
